src/main/java/hudson/remoting/PingFailureAnalyzer.java

```diff
@@ -0,0 +1,225 @@
+package hudson.remoting;
+
+import org.jvnet.animal_sniffer.IgnoreJRERequirement;
+
+import java.io.File;
+import java.io.FileNotFoundException;
+import java.io.PrintWriter;
+import java.io.UnsupportedEncodingException;
+import java.lang.management.LockInfo;
+import java.lang.management.ManagementFactory;
+import java.lang.management.MonitorInfo;
+import java.lang.management.ThreadInfo;
+import java.lang.management.ThreadMXBean;
+import java.text.SimpleDateFormat;
+import java.util.Date;
+
+/**
+ * Generates diagnosis information (threaddump + stacktrace) when the pingThread fails
+ *
+ * <p>
+ * Useful to troubleshoot the slave disconnection failures on the agent side.
```

Comment:
@andresrc I think you are talking about http://javadoc.jenkins-ci.org/jenkins/slaves/PingFailureAnalyzer.html and I am already using it on my code. So I guess, this might be useful for master troubleshooting as well.